mobilecoind-dev-faucet/src/worker.rs

```diff
@@ -0,0 +1,632 @@
+                    // we should drop it from our records.
+                    is_tx_still_in_flight(client, &resp, "Faucet", logger)
+                } else {
+                    // The oneshot receiver resolved in an error, this means, the other side dropped
+                    // this channel, without reporting a SubmitTxResponse. This
+                    // means there was an error building or submitting the Tx,
+                    // and the other side has now dropped this UnspentTxOut. We should requeue it
+                    // so that it can be eventually be spent, and for now we should just purge it.
+                    false
+                }
+            } else {
+                // Still in the queue as far as we know
+                true
+            }
+        });
+
+        // Now, get a fresh unspent tx out list associated to this token
+        let mut resp = {
+            let mut req = mc_mobilecoind_api::GetUnspentTxOutListRequest::new();
+            req.token_id = *self.token_id;
+            req.monitor_id = monitor_id.to_vec();
+
+            client.get_unspent_tx_out_list(&req).map_err(|err| {
+                format!(
+                    "Could not get unspent txout list for token id = {}: {}",
+                    self.token_id, err
+                )
+            })?
+        };
+
+        // Now, check all the reported utxos.
+        // If it is new and has the target value, then queue it
+        let mut output_list_key_images = HashSet::<KeyImage>::default();
+
+        for utxo in resp.output_list.iter() {
+            // Sanity check the token id
+            if utxo.token_id != self.token_id {
+                continue;
+            }
+
+            // Only utxos of exactly the target value are elligible to go in the queue.
+            // The others are "non-target-value utxos" which are candidates to be used
+            // (later) in split transactions that produce more target-value utxos,
+            // if the queue is getting empty.
+            if utxo.value != self.target_value {
+                continue;
+            }
+
+            let key_image: KeyImage = utxo
+                .get_key_image()
+                .try_into()
+                .map_err(|err| format!("invalid key image: {}", err))?;
+            if let Entry::Vacant(e) = self.known_utxos.entry(key_image) {
+                // We found a utxo not in the cache, let's queue and add to cache
+                log::trace!(
+                    logger,
+                    "Queueing a utxo: key_image = {:?}, value = {}",
+                    key_image,
+                    utxo.value
+                );
+                let (tracker, record) = UtxoTracker::new(utxo.clone());
+                // Add to queue depth before push, because we subtract after pop
+                self.queue_depth.fetch_add(1, Ordering::SeqCst);
+                let _ = self.sender.send(record);
```

Comment:
I guess I'm not sure if it's interesting to try to handle this error

This error occurs if we try to add a utxo to the queue, but we cannot. The queue is an unbounded queue, so according to docs it only happens if `close` is called or if the receiver end is dropped.

Close is not called, so the only way this can happen is if the Worker handle struct is dropped. But drop for that function calls join for the handle to this thread. So this really should never happen.

If the receiver end of the queue is dropped, the worker thread cannot do anything useful anymore. If we make this error be map err and returned to caller, then the error will just be logged infinitely in a loop. It might be reasonable to try to say, if this error is encountered, we should try to exit this thread gracefully. But this will also be a lot of plumbing, and I'm not sure it really improves the overall health of the program, because the faucet will stop working if the worker has joined, and the faucet doesn't currently have any way to restart the worker.

Actually I guess the most sensible thing to do is panic, since the worker thread expects this not to ever fail, and the worker thread can't really do anything useful anymore if the queue is closed. I would be open to changing this to a panic

Reply:
Panicking for this exceptional case SGTM